### ISSUE TITLE
improve bios raid handling on system disk

### ISSUE THREAD

**Issue by phillxnet** [rockstor/rockstor-core#1151]
When using bios raid, such as Intel Raid Storage Technology on the system disk there is no system disk identification due to differences in the device naming scheme, ie /dev/md126p3 rather than the expected /dev/sda3. This causes the base md126 as well as the swap partition to be listed in the disks page all with inappropriate flags.

This has been seen in the forum in the following post:-
http://forum.rockstor.com/t/issues-cant-wipe-some-drives-no-smart-diagnostics-disk-serial-numbers-not-legitamate/820/11

I have reproduced the above system disk layout locally:-
![3 8-11-default-install-irst](https://cloud.githubusercontent.com/assets/2521585/12853674/506d0782-cc2d-11e5-8df9-58b3cc96ea6b.png)

**Comment by phillxnet** [comment]
I am having a quick look at this issue currently.

**Comment by phillxnet** [comment]
Not there yet but Flags looking better and root identified and base root dev not showing but.
Outstanding issue is swap partition still showing up and not sure why.
![intel-raid-partly-accomodated](https://cloud.githubusercontent.com/assets/2521585/12858564/c293b8f4-cc48-11e5-9edf-3c9290176c93.png)

Will try and continue with this tomorrow (UK).

**Comment by phillxnet** [comment]
There may also be an opportunity to use the md devices uuid in lieu of our device serial number but will have a look at that when the above issue is sorted.

**Comment by phillxnet** [comment]
Progress so far: the above patches now more elegantly deal with the same Intel Raid Storage bios boot device used as the system disk:-
![intel-raid-accomodated](https://cloud.githubusercontent.com/assets/2521585/12868991/14e3b718-cd0b-11e5-8fee-b929fefd1849.png)

I am now planning to test the same patches for compatibility with a multi md device install such as that included in the docs:-  http://rockstor.com/docs/mdraid-mirror/boot_drive_howto.html which is an all software raid arrangement with no bios raid component.

**Comment by phillxnet** [comment]
I hope to continue with this issue tomorrow (UK).

**Comment by phillxnet** [comment]
When installing according to the docs howto for mirrored os drives via triple md devices ie /boot / swap, the current 3.8-11.07 presents the drives as follows:-
![md-howto-3 8-11 07](https://cloud.githubusercontent.com/assets/2521585/12872935/c653b1a4-cda8-11e5-8570-f07b5a6dad00.png)

**Comment by phillxnet** [comment]
The following is the result of these patches when used with the non btrfs root md howto install:-
![howto-md-results](https://cloud.githubusercontent.com/assets/2521585/12873948/1cdc34a6-cdc2-11e5-859a-a55d89a11f64.png)
I believe this is as expected due to assumptions of root being btrfs and in a partition. It did however lead to a cleaner root_disk reporting implementation which can now deal with root directly on an md device.

**Comment by phillxnet** [comment]
Remaining problems:- 
Need to address properly returning root=True on root md126p3 type device.
Debug and test new get_md_members function.
Test with device that requires get_disk_serial that isn't an md device (ie sdcard) as did minor change to re.match test.

Hope to pick this up again tomorrow.

**Comment by phillxnet** [comment]
Current state of this patch set with the same initial bios raid install as at the top of this issue:-
![bios-md-working-with-model-info](https://cloud.githubusercontent.com/assets/2521585/12898748/23c6feb0-cea6-11e5-947b-bc064689da14.png)
Now properly ascribes root=True in info for root partition md126p3.
ToDo:-
Test get_disk_serial on non md device that requires udev fail over mechanism (ie sdcard).
Possibly use serial numbers with or instead of dev names in raid model info.
Further testing.
Remove issue related logging.

Hope to continue this tomorrow.

**Comment by phillxnet** [comment]
If I leave scope pretty much as is then the resulting pull request from this issue should be available soon.

**Comment by schakrava** [comment]
I skimmed through your comments and some changes, nice work @phillxnet. thanks! looking forward to testing and merging the upcoming pr. good luck

**Comment by phillxnet** [comment]
To address the issue of raid member disks having the cog icon and a popup suggesting they be wiped we could add an extension to the Disk model in:-
src/rockstor/storageadmin/models/disk.py
This extension could be the addition of a role filed such as has just been added to the pool model in:-
https://github.com/rockstor/rockstor-core/commit/9bc97953e390c9687c0a02b7b4dc0142b32f4433
and then the js view:-
src/rockstor/storageadmin/static/storageadmin/js/views/disks.js
could be informed by this role to display a different icon which in turn links to a details page of that device from an md perspective, likewise the same role mechanism could be ascribed (by scan_disks) to the resulting md126p3 device to access it's mdview.  The mouse over popup hint could give a brief description, ie lsblk gives the sda device an FSTYPE="isw_raid_member" this could probably go straight into the disk role.

This additional role field could also accommodate the labelling of devices such as external temp connected USB disks, unique by their serial and their disk model "role" would return upon their re-connection. See proposed feature in issue #1157 .

I think this disk model addition is beyond the scope of this "Part 1" of "improve bios raid handling on system disk" issue so I will leave these cogs on raid members dangling for the time being so the disk model discussion can take place.

**Comment by phillxnet** [comment]
Serial retrieval that uses udev fail over tested as still working after changed. or virtio and mmcblk devs.
Changed over and tested md model info to reference serials of devs listed as members by udevadm. This way one can see which drives were members and we don't have the confusion caused by dev names moving. udevadm info name gives current dev names so we can use them I believe but because we store this info in Disk model it should not be ephemeral, plus if device down or missing then only it's serial will be visible in disks page.

The following shows the same initial bios raid install but having been updated and the 3 changed files overwritten in place. Disk refresh then indicates disks interpretation.
Note the correct labelling of the external usb disk used on another Rockstor with multiple shares that is correctly labelled as btrfs and importable.
![bios-md-serial-with-import](https://cloud.githubusercontent.com/assets/2521585/12924607/d4de224e-cf52-11e5-8f45-dcb9743f331f.png)

See previous comment in this issue for dangling cogs on bios raid members.

**Comment by phillxnet** [comment]
And the same system having just imported from the usb disk:-
![bios-md-serial-after-import](https://cloud.githubusercontent.com/assets/2521585/12924956/e3bca83e-cf53-11e5-84d9-e759e43ad0fb.png)

**Comment by phillxnet** [comment]
The same patches applied to a non md system with a ready to use vda (ie blank) and a FAT32 single partition sdc flagged as expected:-
![non-md-with-md-patches](https://cloud.githubusercontent.com/assets/2521585/12926473/79d32e5e-cf5b-11e5-9c01-23aa5db17b3a.png)
```
sda      8:0    0    8G  0 disk 
├─sda1   8:1    0  500M  0 part /boot
├─sda2   8:2    0  820M  0 part [SWAP]
└─sda3   8:3    0  6.7G  0 part /mnt2/rockstor_rockstor
sdb      8:16   0   10G  0 disk /mnt2/rock-pool
sdc      8:32   0    2G  0 disk 
└─sdc1   8:33   0    2G  0 part 
sr0     11:0    1 1024M  0 rom  
vda    253:0    0    2G  0 disk 
```

**Comment by schakrava** [comment]
Fixed by #1158
